components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/mediators/ClaimBasedResourceAccessValidationMediator.java

```diff
@@ -0,0 +1,142 @@
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * This mediator checks for a value of a pre-configured claim which should be sent in the
+ * JWT access token. This claim value is also configured in the policy which is uploaded to each API resource
+ * in the API Manager publisher. For the mediator to return true, the claim name and the claim value sent in the JWT
+ * access token must be identical to the configured claim name and the value. There is also an option to configure a
+ * so the claim values will be matched against that regex. If one of those are not identical,
+ * the mediator will return false. The value sent in the token claim must be equal or match with the optionally
+ * provided regex. If the "shouldAllowValidation" is true, the flow will be allowed if the claim values are matched. If
+ * it is false, the flow will be allowed if the claim values are not matched.
+ */
+public class ClaimBasedResourceAccessValidationMediator extends AbstractMediator {
+
+    private static final Log log = LogFactory.getLog(ClaimBasedResourceAccessValidationMediator.class);
+    private String accessVerificationClaim;
+    private String accessVerificationClaimValue;
+    private String accessVerificationClaimValueRegex;
+    private boolean shouldAllowValidation;
+    public static final String CLAIMS_MISMATCH_ERROR_MSG = "Configured claim and claim " +
+            "sent in token do not match.";
+
+    @Override
+    public boolean mediate(MessageContext messageContext) {
+
+        String claimValueSentInToken;
+        Map<String, String> jwtTokenClaims = (Map<String, String>) messageContext
+                .getProperty(APIMgtGatewayConstants.JWT_CLAIMS);
+
+        claimValueSentInToken = jwtTokenClaims.get(accessVerificationClaim);
+
+        if (StringUtils.isBlank(claimValueSentInToken)) {
+            log.error("The configured resource access validation claim is " +
+                    "not present in the token.");
+            handleFailure(HttpStatus.SC_FORBIDDEN, messageContext, String.format("Token doesn't contain the " +
+                    "claim \"%s\"", accessVerificationClaim), null);
+            return false;
+        }
+
+        if (StringUtils.isNotBlank(accessVerificationClaimValueRegex)) {
+            log.debug("A regex is provided, hence, validating the claim values using the provided regex.");
+            Pattern pattern = Pattern.compile(accessVerificationClaimValueRegex);
+            Matcher configuredClaimValueMatcher = pattern.matcher(accessVerificationClaimValue);
+            Matcher tokenSentClaimValueMatcher = pattern.matcher(claimValueSentInToken);
+
+            if ((configuredClaimValueMatcher.matches() && tokenSentClaimValueMatcher.matches())
+                    || (!(configuredClaimValueMatcher.matches() && tokenSentClaimValueMatcher.matches())
```

Comment:
Shall we remove the "(!(configuredClaimValueMatcher.matches() && tokenSentClaimValueMatcher.matches())" part since we can directly add the "shouldAllowValidation" after the "||"?

Reply:
Removed with https://github.com/wso2/carbon-apimgt/pull/12247/commits/4662eca21cf9e0d528459b638a49782064d94739.